### PR TITLE
Import scripts from Hoa/Core

### PR DESCRIPTION
**Pull request by jubianchi**
Everything seems OK to me except two things:
- I have to finish the `README.md`
- One thing does not work: I installed Hoa/Cli and its dependency this way:

```
.
├──  Cli/
├──  Console/
├──  Core/
├──  Dispatcher/
├──  Router/
├──  Stream/
└──  UString/
```

From here, when I run `Cli/Bin/hoa` I get the following error:

```
PHP Fatal error:  Class 'Hoa\Router\Cli' not found in Cli/Bin/Hoa.php on line 76
```

I'm trying to simulate a global install of Hoa. Any clue on how to fix this ?

Everything else works (when using composer).

### CODE REVIEW

File: Bin/Hoa.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        'autoload.php'
+    ];
+    foreach ($composer as $path) {
+        if (file_exists($path)) {
+            require_once $path;
+            break;
+        }
+    }
+    if (!defined('HOA')) {
+        require_once dirname(__DIR__) . DIRECTORY_SEPARATOR .
```
Comment (jubianchi) [inline Bin/Hoa.php]: @Hywan You were right here: no need to loop through a list of files to load `Hoa/Core`: if we reach this block it means we are not using composer (composer automatically loads `Hoa/Core` with `autoload.files`) so there is a single possible path to load.

**Comment by Hywan** [comment]
It is `Ustring`, not `UString` (because Composer puts everything in lowercase so cannot introduce specific uppercase letter). Retry with this.

File: Bin/Hoa.php (hunk 1 above)
Comment (Hywan) [inline Bin/Hoa.php]: CS:

``` php
require_once
    dirname(__DIR__) . DIRECTORY_SEPARATOR .
    '..' . DIRE …
    …
    'Core.php';
```

File: Bin/Hoa.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            '..' . DIRECTORY_SEPARATOR .
+            'Core' . DIRECTORY_SEPARATOR .
+            'Core.php';
+    }
+}
+Core::enableErrorHandler();
+Core::enableExceptionHandler();
```
Comment (Hywan) [inline Bin/Hoa.php:74]: CS: An empty line after `Core::enableExceptionHandler();` (before the comment).

**Comment by jubianchi** [comment]
@Hywan same error. the errors comes from `Hoa/Router` ;)

File: Bin/Hoa.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    ]);
+    $dispatcher->setKitName('Hoa\Console\Dispatcher\Kit');
+    exit((int) $dispatcher->dispatch($router));
+} catch (Core\Exception $e) {
+    $message = $e->raise(true);
+    $code    = 1;
+} catch (\Exception $e) {
+    $message = $e->getMessage();
+    $code    = 2;
+}
```
Comment (Hywan) [inline Bin/Hoa.php:106]: CS: An empty line after `}`.

File: Bin/hoa [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
+ * @author     Ivan Enderlin <ivan.enderlin@hoa-project.net>
```
Comment (Hywan) [inline Bin/hoa]: Remove this line.

File: Bin/Hoa.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (Hywan) [inline Bin/Hoa.php:34]: CS: New line.

File: Bin/Hoa.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+use Hoa\Console;
+use Hoa\Core;
+use Hoa\Dispatcher;
+use Hoa\Router;
```
Comment (Hywan) [inline Bin/Hoa.php:39]: CS: New line.

File: Bin/hoa [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ * @copyright  Copyright © 2007-2015 Ivan Enderlin.
```
Comment (Hywan) [inline Bin/hoa]: Replace by `@copyright  Copyright © 2007-2015 Hoa community`.

File: Bin/hoa.bat [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+REM * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+REM * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+REM * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+REM * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+REM * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+REM * POSSIBILITY OF SUCH DAMAGE.
+REM **
+REM
+REM **
+REM * @author     Ivan Enderlin <ivan.enderlin@hoa-project.net>
```
Comment (Hywan) [inline Bin/hoa.bat]: Remove this line.

File: Bin/hoa.bat [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+REM * @copyright  Copyright © 2007-2015 Ivan Enderlin.
```
Comment (Hywan) [inline Bin/hoa.bat]: Replace by `@copyright  Copyright © 2007-2015 Hoa community`.

File: CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,3 @@
+# `dev-master` / `1.x-dev`
```
Comment (Hywan) [inline CHANGELOG.md]: Remove the whole file. It's going to be automatically generated.

File: composer.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            "homepage": "http://hoa-project.net/"
+        }
+    ],
+    "support": {
+        "email" : "support@lists.hoa-project.net",
+        "irc"   : "irc://irc.freenode.org/hoaproject",
+        "source": "http://git.hoa-project.net/"
+    },
+    "require": {
+        "hoa/core"      : "~1.0 || ~2.0",
```
Comment (Hywan) [inline composer.json]: Use just the highest master compatibility number. So `"hoa/core": "~2.0"` for instance.

**Comment by Hywan** [comment]
Did you just copy/paste the files? There is some issues with the CS.

File: CHANGELOG.md (hunk 10 above)
Comment (jubianchi) [inline CHANGELOG.md]: nice!

**Comment by jubianchi** [comment]
@Hywan I just took the files from `Hoa/Core` and there were some CS issue there but no problem, I fixed them all ;)

**Comment by Hywan** [comment]
@jubianchi: About `Hoa\Router`, how did you install the libraries? I would like to try to reproduce.

**Comment by Hywan** [comment]
@jubianchi: Oh, good!

**Comment by jubianchi** [comment]
> @jubianchi: About Hoa\Router, how did you install the libraries? I would like to try to reproduce.

@Hywan I cloned the required repositories manually in a directory.

**Comment by Hywan** [comment]
@jubianchi: I will try as soon as possible.

**Comment by jubianchi** [comment]
@Hywan thanks, let me know if I have to change anything else ;)

I'm going to take a nap, I'm sick :mask:

**Comment by Hywan** [comment]
@jubianchi Rename your parent directory `Hoa` and it will work ;-).

**Comment by Hywan** [comment]
@jubianchi I am sick too… Damn, firewall was off?

File: Bin/Hoa.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        ]
+    );
+    $dispatcher = new Dispatcher\ClassMethod([
+        'synchronous.call'
+            => '(:%variables.vendor:lU:)\(:%variables.library:lU:)\Bin\(:%variables.command:lU:)',
+        'synchronous.able'
+            => 'main'
+    ]);
+    $dispatcher->setKitName('Hoa\Console\Dispatcher\Kit');
+    exit((int) $dispatcher->dispatch($router));
```
Comment (Hywan) [inline Bin/Hoa.php:99]: CS: New line before `exit` (sorry).

File: Bin/Hoa.php [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+ob_start();
+Console\Cursor::colorize('foreground(white) background(red)');
+echo $message, "\n";
+Console\Cursor::colorize('normal');
+$content = ob_get_contents();
+ob_end_clean();
+file_put_contents('php://stderr', $content);
+exit($code);
```
Comment (Hywan) [inline Bin/Hoa.php:115]: CS: New line before `exit` (sorry).

File: Bin/hoa [hunk 14]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env php
+<?php
```
Comment (Hywan) [inline Bin/hoa:2]: CS: New line here.

File: Bin/hoa [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (Hywan) [inline Bin/hoa:35]: CS: Empty line here.

File: Bin/hoa [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * @copyright  Copyright © 2007-2015 Hoa community
+ * @license    New BSD License
+ */
```
Comment (Hywan) [inline Bin/hoa:39]: CS: Empty line here.

**Comment by Hywan** [comment]
Thank you for your PR!

**Comment by Hywan** [comment]
https://waffle.io/hoaproject/Cli for the overview of this repository.

**Pull request merged**